### PR TITLE
Plot State City Visualization Fixes

### PR DESCRIPTION
**Pull request by AlexanderGroeger**
This is my first PR, please correct me as needed. :)

### Summary
The `qiskit.visualization` module has a function `plot_state_city` which yields a matplotlib 3d bar chart representing the density matrix of a quantum state. 

Fixes https://github.com/Qiskit/qiskit/issues/8413
Fixes https://github.com/Qiskit/qiskit/issues/3315
Related to https://github.com/Qiskit/qiskit/pull/10862

### Details and comments
Note for some extreme choices of `figsize`, the plot labels can still be cut off. The primary issue is matplotlib cannot use `tight_layout` because there's a lack of space.

### CODE REVIEW

**Comment by AlexanderGroeger** [comment]
Looks like the qiskit-terra test failed because `test_plot_state_city` did not succeed. Apparently the error was `AssertionError: 0.930396 not greater than or equal to 0.99` coming from 
```
ratio = VisualTestUtilities._save_diff(
            self._image_path(fname),
            self._reference_path(fname),
            fname,
            FAILURE_DIFF_DIR,
            FAILURE_PREFIX,
        )
        self.assertGreaterEqual(ratio, 0.99)
```

**Comment by Cryoris** [comment]
Since your PR changes the output of the `plot_state_city` function, you probably have to change the reference pictures in used in the tests. See e.g. the comments of #10619 on how to do this 🙂

**Comment by AlexanderGroeger** [comment]
I believe this PR should be ready for review, sorry about the multiple merges; I wasn't sure if I should keep my PR up-to-date or not.

**Comment by AlexanderGroeger** [comment]
If this PR could be reviewed and closed soon, that would be great. Thanks!

**Comment by 1ucian0** [comment]
Thanks a lot for your contribution. Original, the bug in https://github.com/Qiskit/qiskit/pull/10862 was introduced by [a rounding error that made testing](https://github.com/Qiskit/qiskit/pull/9961/commits/e6528cfddb7df82b1d9f83b34536fb220d3e1e4b#r1222952369). Do you know if your changes are sensitive to them?

I never understood why there is a color difference for 0.0:
state_city

The RGB values shouldn't be the same?

**Comment by AlexanderGroeger** [comment]
Regarding [a rounding error that made testing](https://github.com/Qiskit/qiskit/pull/9961/commits/e6528cfddb7df82b1d9f83b34536fb220d3e1e4b#r1222952369), I removed this line altogether without using any substitute code. I didn't quite understand why this was an issue. It sounded like bars with zero height were considered negative in the visualization, but I'm wondering if this was just a problem with how the dividing zero plane was plotted in which that has been corrected.
The reason for the color difference eludes me too. I was initially considering the issue was with [generate_facecolors](https://github.com/AlexanderGroeger/qiskit-terra/blob/9c3bfcd9b59e586c0f789133fb701500602510c3/qiskit/visualization/state_visualization.py#L1085C5-L1085C24), but there is nothing obvious there. I'm starting to wonder if it related to how matplotlib handles 3d projections.

File: qiskit/visualization/state_visualization.py [hunk 1]
```diff
@@ -511,109 +510,105 @@ def plot_state_city(
...
-            else:
-                zorder = 0
-            b1 = ax1.bar3d(
-                xpos[idx],
-                ypos[idx],
-                cur_zpos,
-                dx[idx],
-                dy[idx],
-                dzr[idx],
+    print(real_color, imag_color)
```
Comment (1ucian0) [inline qiskit/visualization/state_visualization.py]: this looks like debugging print

**Comment by 1ucian0** [comment]
Is this fixing https://github.com/Qiskit/qiskit-terra/issues/8413 ,  https://github.com/Qiskit/qiskit-terra/issues/3315 and https://github.com/Qiskit/qiskit/pull/10862 ? If so uses `Fixes #<number>` in the main message for close them automatically once this gets merged.

File: qiskit/visualization/state_visualization.py (hunk 1 above)
Comment (AlexanderGroeger) [reply qiskit/visualization/state_visualization.py]: Yes it was. I have made a commit [87ca633](https://github.com/Qiskit/qiskit/pull/10590/commits/87ca63330f27f3a3e731fc94aa61a0c56f71cf78) to remove it. Thank you!

**Comment by AlexanderGroeger** [comment]
> Is this fixing #8413 , #3315 and #10862 ? If so uses `Fixes #<number>` in the main message for close them automatically once this gets merged.

Yes. I edited the PR description to reflect this. I'm still a bit new to some of these expectations, so thank you!

**Comment by 1ucian0** [comment]
Take a look to the black issue.

Dropping the label for the axes (`Re` and `Im`) was intentional?

**Comment by AlexanderGroeger** [comment]
Yes. I figured putting subtitles over each subplot would be easier to read instead. The primarily reason why the y-axis labels were moved was because of spacing issues related to https://github.com/Qiskit/qiskit/issues/8413.

File: qiskit/visualization/state_visualization.py [hunk 2]
```diff
@@ -511,109 +510,103 @@ def plot_state_city(
...
-        if min(dzi) < 0 < max(dzi):
-            ax2.add_collection3d(pc2)
-        ax2.set_xticks(np.arange(0.5, lx + 0.5, 1))
-        ax2.set_yticks(np.arange(0.5, ly + 0.5, 1))
-        if min_dzi != max_dzi:
-            eps = 0
-            ax2.axes.set_zlim3d(np.min(dzi), max(np.max(dzr) + 1e-9, np.max(dzi) + eps))
+            positive_bars.set_facecolor(fc)
+
+        ax.set_title(f"{zlabel} Amplitude", fontsize=max_font_size)
```
Comment (1ucian0) [inline qiskit/visualization/state_visualization.py]:
```suggestion
        ax.set_title(f"{zlabel} Amplitude (ρ)", fontsize=max_font_size)
```

**Review by 1ucian0 — APPROVED** [review]
Thanks!

**Pull request merged**